src/bucket/LiveBucket.cpp

```diff
@@ -450,6 +458,158 @@ LiveBucket::maybeInitializeCache(size_t totalBucketListAccountsSizeBytes,
+            LiveBucket::countNewEntryType(mc, newEntries[newIdx]);
+            mergedEntries.push_back(newEntries[newIdx]);
+            ++newIdx;
+        }
+        else
+        {
+            releaseAssert(oldIdx < oldEntries.size() &&
+                          newIdx < newEntries.size());
+
+            // Entries have equal keys, handle the special merge cases
+            auto const& oldEntry = oldEntries[oldIdx];
+            auto const& newEntry = newEntries[newIdx];
+            LiveBucket::checkProtocolLegality(oldEntry, maxProtocolVersion);
+            LiveBucket::checkProtocolLegality(newEntry, maxProtocolVersion);
+            LiveBucket::countOldEntryType(mc, oldEntry);
+            LiveBucket::countNewEntryType(mc, newEntry);
+
+            if (newEntry.type() == INITENTRY)
+            {
+                // The only legal new-is-INIT case is merging a delete+create to
+                // an update.
+                if (oldEntry.type() != DEADENTRY)
+                {
+                    throw std::runtime_error(
+                        "Malformed bucket: old non-DEAD + new INIT.");
+                }
+
+                BucketEntry newLive;
+                newLive.type(LIVEENTRY);
+                newLive.liveEntry() = newEntry.liveEntry();
+                ++mc.mNewInitEntriesMergedWithOldDead;
+                mergedEntries.push_back(newLive);
+            }
+            else if (oldEntry.type() == INITENTRY)
+            {
+                // If we get here, new is not INIT; may be LIVE or DEAD.
+                if (newEntry.type() == LIVEENTRY)
+                {
+                    // Merge a create+update to a fresher create.
+                    BucketEntry newInit;
+                    newInit.type(INITENTRY);
+                    newInit.liveEntry() = newEntry.liveEntry();
+                    ++mc.mOldInitEntriesMergedWithNewLive;
+                    mergedEntries.push_back(newInit);
+                }
+                else
+                {
+                    // Merge a create+delete to nothingness.
+                    ++mc.mOldInitEntriesMergedWithNewDead;
+                }
+            }
+            else
+            {
+                // Neither is in INIT state, take the newer one.
+                ++mc.mNewEntriesMergedWithOldNeitherInit;
+                mergedEntries.push_back(newEntry);
+            }
+
+            ++oldIdx;
+            ++newIdx;
+        }
+    }
+
+    if (countMergeEvents)
+    {
+        bucketManager.incrMergeCounters(mc);
+    }
+
+    // Write merge output to a bucket and save to disk
+    LiveBucketOutputIterator out(bucketManager.getTmpDir(),
+                                 /*keepTombstoneEntries=*/true, meta, mc, ctx,
+                                 doFsync);
+
+    for (auto const& e : mergedEntries)
+    {
+        out.put(e);
+    }
+
+    // Store the merged entries in memory in the new bucket in case this is
```

Comment:
I'm a bit confused about this comment: why do we merge level 1 in-memory?

Reply:
Ahh, this comment is wrong and I need to correct it. How this should read is

```
    // Store the merged entries in memory in the new bucket in case this is
    // still a level 0 curr bucket. If the Bucket is a level 0 snap, this won't be used but
    // will be GC'd quickly anyway
```